## 1. Data quality 
a. uniqueness (pelanggaran : duplikat) -> kita cek bahwa tidak ada data yang duplikat. contoh : primary key muncul beberapa kali \
b. completeness -> datanya ada yang null atau tidak \
c. accuracy -> value nya itu dalam range yang diperbolehkan atau value real \
d. consistancy -> penulisan nya sama atau tidak / jenisnya seragam atau tidak

In [1]:
import pandas as pd

In [2]:
# extract data
df = pd.read_csv('C:/Users/irfan/OneDrive/Documents/project/bi_bootcamp_binar/BIM wave 3/dataset/ecommerce_women_clothing.csv')

In [3]:
df.head()

order_id          order_date   sku      color      size  unitPrice  \
0         1   2022/6/1 16:05:00   708  Dark Blue       2XL        298   
1         1   2022/6/1 16:05:00    89  Dark Blue       2XL        258   
2         2  2022/6/10 17:03:00  bobo      Cream  One Size        228   
3         3  2022/6/10 19:59:00   799  Dark Blue        XL        264   
4         4  2022/6/11 11:42:00   799  Dark Blue         M        264   

   quantity  REVENUE                                 temp_value  
0         1      298    144713.6701388889708Dark Blue2XL2981298  
1         1      258     144713.670138888989Dark Blue2XL2581258  
2         1      228  244722.7104166667boboCreamOne Size2281228  
3         1      264     344722.8326388889799Dark BlueXL2641264  
4         1      264            444723.4875799Dark BlueM2641264

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 528 entries, 0 to 527
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   order_id    528 non-null    int64 
 1   order_date  528 non-null    object
 2   sku         528 non-null    object
 3   color       528 non-null    object
 4   size        492 non-null    object
 5   unitPrice   528 non-null    int64 
 6   quantity    528 non-null    int64 
 7   REVENUE     528 non-null    int64 
 8   temp_value  528 non-null    object
dtypes: int64(4), object(5)
memory usage: 37.2+ KB


### a.uniqueness

In [4]:
# cek duplikat
# template -> df.duplicated() -> cek duplication
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
523    False
524    False
525    False
526    False
527    False
Length: 528, dtype: bool

In [5]:
# jumlah row yang duplicate
sum(df.duplicated())

1

In [6]:
# memunculkan data yang duplicate
# template -> nama_va[nama_var.duplicated()]
df[df.duplicated()]

order_id          order_date  sku      color size  unitPrice  quantity  \
396       201  2022/8/15 12:38:00  799  Dark Blue   XL        266         1   

     REVENUE                                temp_value  
396      266  20144788.5263888889799Dark BlueXL2661266

Terdapat duplicate data sejumlah 1 data, dengan order_id = 201

In [10]:
df[['order_id', 'order_date']].groupby(by = 'order_id').count()
# order_id bukan primary key

order_date
order_id            
1                  3
2                  2
3                  2
4                  2
5                  2
...              ...
269                2
270                1
271                1
272                1
273                1

[273 rows x 1 columns]

In [12]:
# contoh checking duplicate untuk primary key
df[df.duplicated('order_id')]

order_id          order_date   sku       color      size  unitPrice  \
1           1   2022/6/1 16:05:00    89   Dark Blue       2XL        258   
16         15  2022/6/14 14:38:00  bobo        Blue  One Size        228   
17         15  2022/6/14 14:38:00  bobo  Light Gray  One Size        228   
21         18  2022/6/15 13:36:00    77           C       NaN        258   
22         18  2022/6/15 13:36:00    77           D       NaN        258   
..        ...                 ...   ...         ...       ...        ...   
523       171    2022/9/9 6:27:00   799   Dark Blue         M        265   
524       172    2022/9/9 7:18:00   708   Dark Blue       4XL        278   
525       173    2022/9/9 7:22:00   799   Dark Blue         M        265   
526       174    2022/9/9 8:25:00   799   Dark Blue         L        265   
527       175    2022/9/9 9:47:00   799   Dark Blue        XL        265   

     quantity  REVENUE                                       temp_value  
1           1      258           144713.670138888989Dark Blue2XL2581258  
16          1      228        1544726.6097222222boboBlueOne Size2281228  
17          1      228  1544726.6097222222boboLight GrayOne Size2281228  
21          1      258                     1844727.566666666777C2581258  
22          1      258                     1844727.566666666777D2581258  
..        ...      ...                                              ...  
523         1      265               17144813.26875799Dark BlueM2651265  
524         1      278        17244813.3041666667708Dark Blue4XL2781278  
525         1      265          17344813.3069444444799Dark BlueM2651265  
526         1      265          17444813.3506944444799Dark BlueL2651265  
527         1      265         17544813.4076388889799Dark BlueXL2651265  

[255 rows x 9 columns]

### b. completeness

In [13]:
# template -> nama_var.isna().any()
df.isna().any()
# ada value null pada kolom size, dapat dilihat dari hasil balikan True

order_id      False
order_date    False
sku           False
color         False
size           True
unitPrice     False
quantity      False
REVENUE       False
temp_value    False
dtype: bool

In [14]:
# check null values in a column
# template -> nama_var['column'].isna()
df['size'].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
523    False
524    False
525    False
526    False
527    False
Name: size, Length: 528, dtype: bool

In [15]:
# count total null rows
sum(df['size'].isna())

36

In [16]:
df[df['size'].isna()]

order_id          order_date  sku      color size  unitPrice  quantity  \
20         18  2022/6/15 13:36:00   77          B  NaN        258         1   
21         18  2022/6/15 13:36:00   77          C  NaN        258         1   
22         18  2022/6/15 13:36:00   77          D  NaN        258         1   
23         18  2022/6/15 13:36:00   77          E  NaN        258         1   
24         18  2022/6/15 13:36:00   77          A  NaN        288         1   
66         55  2022/6/30 12:26:00  439       Rust  NaN        258         1   
68         57    2022/6/5 7:18:00   77          B  NaN        258         1   
81         69    2022/6/6 3:56:00   77          B  NaN        258         1   
82         69    2022/6/6 3:56:00   77          D  NaN        258         1   
83         69    2022/6/6 3:56:00   77          A  NaN        288         1   
137       115  2022/7/21 17:56:00   77          C  NaN        239         1   
193       167   2022/7/7 11:56:00   77          B  NaN        239         1   
206       178   2022/7/9 21:20:00   77          A  NaN        239         1   
233       204  2022/8/15 19:15:00   77          C  NaN        239         1   
234       204  2022/8/15 19:15:00  239      Green  NaN        245         1   
235       204  2022/8/15 19:15:00  239      White  NaN        245         1   
236       204  2022/8/15 19:15:00  239    Mustard  NaN        245         1   
237       204  2022/8/15 19:15:00  239       Rust  NaN        245         1   
238       204  2022/8/15 19:15:00  339      Brown  NaN        254         1   
239       204  2022/8/15 19:15:00  339      Green  NaN        254         1   
282       244  2022/8/25 22:07:00  239       Gray  NaN        278         1   
325        11   2022/8/4 12:30:00   77          B  NaN        239         1   
326        11   2022/8/4 12:30:00   77          C  NaN        239         1   
335        17   2022/8/6 11:04:00  339  Navy Blue  NaN        288         1   
336        17   2022/8/6 11:04:00  339      Mocha  NaN        288         1   
346        25   2022/8/8 15:11:00   77          B  NaN        239         1   
347        25   2022/8/8 15:11:00  239  Navy Blue  NaN        239         1   
348        25   2022/8/8 15:11:00  239      Brown  NaN        239         1   
349        25   2022/8/8 15:11:00  239   Skintone  NaN        278         1   
369        45  2022/9/10 18:43:00  239  Navy Blue  NaN        278         1   
408        75  2022/9/17 21:03:00  239       Rust  NaN        278         1   
415        81   2022/9/18 8:12:00  239     Yellow  NaN        278         1   
444       106  2022/9/24 15:57:00  239       Gray  NaN        259         1   
459       121  2022/9/29 21:14:00  239       Gray  NaN        278         1   
464       125   2022/9/3 22:13:00  239  Navy Blue  NaN        278         1   
490       138   2022/9/4 22:57:00  239      Black  NaN        278         1   

     REVENUE                              temp_value  
20       258            1844727.566666666777B2581258  
21       258            1844727.566666666777C2581258  
22       258            1844727.566666666777D2581258  
23       258            1844727.566666666777E2581258  
24       288            1844727.566666666777A2881288  
66       258        5544742.5180555556439Rust2581258  
68       258            5744717.304166666777B2581258  
81       258            6944718.163888888977B2581258  
82       258            6944718.163888888977D2581258  
83       288            6944718.163888888977A2881288  
137      239           11544763.747222222277C2391239  
193      239           16744749.497222222277B2391239  
206      239           17844751.888888888977A2391239  
233      239           20444788.802083333377C2391239  
234      245      20444788.8020833333239Green2451245  
235      245      20444788.8020833333239White2451245  
236      245    20444788.8020833333239Mustard2451245  
237      245       20444788.8020833333239Rust2451245  
238      254      20444788.802083333

Data null terdapat pada kolom `size` sebanyak 36 kolom

### c. accuracy

In [34]:
df.describe([.25, .5, .75, .98])

order_id     unitPrice    quantity     REVENUE
count  528.000000    528.000000  528.000000  528.000000
mean   115.475379    510.250000    1.956439  278.001894
std     73.132456   4406.015775   21.715951   44.701689
min      1.000000    191.000000    1.000000  191.000000
25%     54.000000    266.000000    1.000000  266.000000
50%    110.000000    278.000000    1.000000  278.000000
75%    166.250000    288.000000    1.000000  288.000000
98%    264.000000    315.000000    1.000000  318.000000
max    273.000000  99999.000000  500.000000  899.000000

In [36]:
df[df['unitPrice'] > 400]

order_id         order_date  sku      color size  unitPrice  quantity  \
94         75  2022/7/1 18:05:00  799  Dark Blue   XL      15100         1   
279       241  2022/8/24 8:05:00  799  Dark Blue   XL      99999         1   
331        15  2022/8/5 19:33:00  708  Dark Blue  4XL        899         1   
505       153  2022/9/8 19:32:00  799  Dark Blue    M      10000         1   

     REVENUE                                  temp_value  
94       288   7544743.7534722222799Dark BlueXL151001288  
279      288  24144797.3368055556799Dark BlueXL999991288  
331      899    1544778.8145833333708Dark Blue4XL8991899  
505      288   15344812.8138888889799Dark BlueM100001288

ada 3 row data yang memiliki unitPrice diatas expected value. Value > 10000, dan anehnya valuenya tidak match dengan revenue

In [37]:
df[df['quantity'] > 2]

order_id         order_date  sku      color size  unitPrice  quantity  \
256       219  2022/8/17 4:12:00  799  Dark Blue    M        266       500   

     REVENUE                          temp_value  
256      266  21944790.175799Dark BlueM266500266

Ada 1 row data yang quantity 500, sedangkan P98 quantity = 1. Dan ketika cek ke revenue angka kurang valid

### d. consistancy

#### - Check column naming consistancy 

In [18]:
df.columns

Index(['order_id', 'order_date', 'sku', 'color', 'size', 'unitPrice',
       'quantity', 'REVENUE', 'temp_value'],
      dtype='object')

- unitPrice : ditulis dalam camelcase, yang lain pake snakecase
- REVENUE : ditulis dalam capital, yang lain huruf kecil

In [20]:
# mengambil unique value dari tiap kolom
# template -> nama_var['column'].unique()
df['order_id'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

Mencari anomali
1. cek dulu unik value \
2. cek berapa jumlah yang tidak sesuai/inconsistant

In [22]:
df['sku'].unique()

array(['708', '89', 'bobo', '799', '239', '77', '897', '9699', '127',
       '1719', '439', '3799', '229', '2499', '79', '29', '1499', '628',
       '61399', '339', '218', '8499', '3081', '539'], dtype=object)

In [25]:
# filter data with sku = bobo
df[df['sku'] == 'bobo']

order_id          order_date   sku       color      size  unitPrice  \
2           2  2022/6/10 17:03:00  bobo       Cream  One Size        228   
15         15  2022/6/14 14:38:00  bobo   Navy Blue  One Size        228   
16         15  2022/6/14 14:38:00  bobo        Blue  One Size        228   
17         15  2022/6/14 14:38:00  bobo  Light Gray  One Size        228   
262       225  2022/8/19 21:42:00  bobo      Matcha  One Size        228   
308       268   2022/8/3 23:03:00  bobo   Navy Blue  One Size        191   
327        12   2022/8/4 21:11:00  bobo      Maroon  One Size        191   
328        12   2022/8/4 21:11:00  bobo        Blue  One Size        191   
343        24    2022/8/8 0:58:00  bobo       Cream  One Size        191   
344        24    2022/8/8 0:58:00  bobo   Navy Blue  One Size        191   
345        24    2022/8/8 0:58:00  bobo        Blue  One Size        191   
416        81   2022/9/18 8:12:00  bobo        Pink  One Size        228   
429        92  2022/9/20 18:15:00  bobo  Red Orange  One Size        228   
430        92  2022/9/20 18:15:00  bobo      Matcha  One Size        228   
467       128  2022/9/30 13:14:00  bobo       White  One Size        206   
518       166   2022/9/9 21:17:00  bobo       White  One Size        228   

     quantity  REVENUE                                       temp_value  
2           1      228        244722.7104166667boboCreamOne Size2281228  
15          1      228   1544726.6097222222boboNavy BlueOne Size2281228  
16          1      228        1544726.6097222222boboBlueOne Size2281228  
17          1      228  1544726.6097222222boboLight GrayOne Size2281228  
262         1      228     22544792.9041666667boboMatchaOne Size2281228  
308         1      191  26844776.9604166667boboNavy BlueOne Size1911191  
327         1      191      1244777.8826388889boboMaroonOne Size1911191  
328         1      191        1244777.8826388889boboBlueOne Size1911191  
343         2      382       2444781.0402777778boboCreamOne Size1912382  
344         1      191   2444781.0402777778boboNavy BlueOne Size1911191  
345         1      191        2444781.0402777778boboBlueOne Size1911191  
416         1      228        8144822.3416666667boboPinkOne Size2281228  
429         1      228  9244824.7604166667boboRed OrangeOne Size2281228  
430         1      228      9244824.7604166667boboMatchaOne Size2281228  
467         1      206      12844834.5513888889boboWhiteOne Size2061206  
518         1      228      16644813.8868055556boboWhiteOne Size2281228

In [27]:
# len sku = bobo
len(df[df['sku'] == 'bobo'])

16

Terdapat data inconsisten pada kolom `sku`  dengan value **bobo** . Jumlahnya = 16

In [28]:
df.columns

Index(['order_id', 'order_date', 'sku', 'color', 'size', 'unitPrice',
       'quantity', 'REVENUE', 'temp_value'],
      dtype='object')

In [29]:
df['color'].unique()

array(['Dark Blue', 'Cream', 'Black', 'Navy Blue', 'Blue', 'Light Gray',
       'B', 'C', 'D', 'E', 'A', 'Mid Wash', 'Light Blue', 'Dark Wash',
       'Thin Blue', 'Rust', 'Grey Wash', 'Thin Black', '922 Black',
       '922 Army green', 'Gray Stripes', '920 Navy blue', 'Light Wash',
       'Blue Stripes', 'Black Stripes', 'Green', 'White', 'Mustard',
       'Brown', 'Matcha', 'Gray', 'Midnight Blue', 'Maroon', 'Mocha',
       'Skintone', 'Cornflower Blue', 'Yellow', 'Pink', 'Red Orange'],
      dtype=object)

In [30]:
df[['order_id','color']].groupby(by='color').count()

order_id
color                    
920 Navy blue           1
922 Army green          1
922 Black               1
A                       3
B                       6
Black                  14
Black Stripes           1
Blue                    4
Blue Stripes            2
Brown                   3
C                       4
Cornflower Blue         1
Cream                   2
D                       2
Dark Blue             398
Dark Wash               5
E                       1
Gray                    3
Gray Stripes            1
Green                   2
Grey Wash               9
Light Blue             26
Light Gray              1
Light Wash              1
Maroon                  1
Matcha                  2
Mid Wash                8
Midnight Blue           2
Mocha                   1
Mustard                 1
Navy Blue               7
Pink                    1
Red Orange              1
Rust                    3
Skintone                1
Thin Black              1
Thin Blue               3
White                   3
Yellow                  1

In [32]:
df['color'].value_counts()

color
Dark Blue          398
Light Blue          26
Black               14
Grey Wash            9
Mid Wash             8
Navy Blue            7
B                    6
Dark Wash            5
Blue                 4
C                    4
Brown                3
White                3
Gray                 3
Rust                 3
Thin Blue            3
A                    3
Cream                2
Midnight Blue        2
Blue Stripes         2
D                    2
Green                2
Matcha               2
Yellow               1
Cornflower Blue      1
Skintone             1
Mocha                1
Maroon               1
Pink                 1
922 Army green       1
Mustard              1
Black Stripes        1
Light Wash           1
920 Navy blue        1
Gray Stripes         1
922 Black            1
Thin Black           1
E                    1
Light Gray           1
Red Orange           1
Name: count, dtype: int64